### PR TITLE
Ms ssim

### PR DESCRIPTION
**Pull request by jennyfolkesson**
I thought this would work by just swapping put ssim for ssim_multiscale pretty much, but it gives me one real loss value in first batch and then just generates nans...
Please let me know if you notice anything obviously wrong here.

### CODE REVIEW

File: micro_dl/train/metrics.py [hunk 1]
```diff
@@ -58,39 +72,70 @@ def dice_coef(y_true, y_pred, smooth=1.):
...
+        difference between the maximum the and minimum allowed values).
+    :return float ms_ssim: Mean SSIM over images in the batch
+    """
+    msssim = K.mean(tf.image.ssim_multiscale(y_true, y_pred, max_val=max_val))
+    # If you're getting nans
+    msssim = K.where(K.is_nan(msssim), 0., msssim)
+    if msssim is None:
+        msssim = K.constant(0.)
+    return msssim
+
```
Comment (anithapk) [inline micro_dl/train/metrics.py]: Hey, I ran into the same issues in the pytorch version and this is how I solved it:
class MS_SSIM(torch.nn.Module):
    def __init__(self, img_max_int, kernel_size = 11, 
                 sigmas=(0.5, 1.0, 2.0, 4.0, 8.0),
                 num_channels=1, 
                 size_average = True, k1=0.01, k2=0.03):
        
        super().__init__()
        self.img_max_int = img_max_int
        assert kernel_size % 2 != 0, 'odd size kernel needed'
        self.kernel_size = kernel_size
        self.sigmas = sigmas
        self.num_channels = num_channels 
        self.size_average = size_average
        self.kernel = create_gauss_filter(kernel_size, self.num_channels)
        self.k1 = k1
        self.k2 = k2
        kernels = []
        for si in sigmas:
            cur_kernel = create_gauss_filter(kernel_size,
                                             sigma=si, 
                                             num_channels=self.num_channels)
            kernels.append(cur_kernel)
        self.kernels = kernels
        
    def forward(self, img1, img2):
        """
        https://stackoverflow.com/questions/39051451/ssim-ms-ssim-for-tensorflow
        https://github.com/ruthcfong/pytorch-msssim/blob/master/pytorch_msssim/__init__.py
        """
        (_, cur_num_channels, _, _) = img1.size()
        # weights = Variable(torch.FloatTensor([0.0448, 0.2856, 0.3001, 0.2363, 0.1333]))
        weights = Variable(torch.FloatTensor([0.25, 0.3, 0.25, 0.1, 0.1]))
        weights = weights.cuda(img1.get_device())
        lm = []
        for idx, si in enumerate(self.sigmas):
            cur_kernel = self.kernels[idx]
            if img1.is_cuda:
                cur_kernel = cur_kernel.cuda(img1.get_device())
            cur_ssim_mu, cur_ssim_sigma = _ssim(
                img1, img2, self.num_channels, 
                self.img_max_int, cur_kernel, 
                self.kernel_size, size_average=False, 
                k1=self.k1, k2=self.k2)
            # normalize ssim to lie b/w 0 and 1 to avoid NaN's when ^wts
            cur_ssim_mu = (cur_ssim_mu + 1) / 2
            cur_ssim_sigma = (cur_ssim_sigma + 1) / 2
            lm.append(cur_ssim_mu)
            if idx == 0:
                cs = cur_ssim_sigma ** weights[idx]
            else:
                cur_ssim_sigma = cur_ssim_sigma ** weights[idx]
                cs *= cur_ssim_sigma
                #cs[torch.isnan(cs)] = 0
        ssim_mu = lm[-1] ** weights[-1]
        if self.size_average:
            ms_ssim = torch.mean(ssim_mu) * torch.mean(cs) 
            return ms_ssim
        if not self.size_average:
            ms_ssim = ssim_mu * cs
            return ms_ssim
Hope this helps!

Comment (jennyfolkesson) [reply micro_dl/train/metrics.py]: @anithapk thank you for the help! I updated the metric to move ground truth and predictions to a positive range before computing ms-ssim to avoid nans.
It seems to be working for training so this PR is ready for review now. Thanks!

File: micro_dl/train/losses.py [hunk 2]
```diff
@@ -23,8 +22,12 @@ def mae_loss(y_true, y_pred, mean_loss=True):
 
 
 def mse_loss(y_true, y_pred, mean_loss=True):
-    """Mean squared loss"""
+    """Mean squared loss
 
+    :param y_true: Ground truth
+    :param y_pred: Prediction
+    :return float: Mean absolute error loss
```
Comment (anithapk) [inline micro_dl/train/losses.py]: **squared**

File: micro_dl/train/metrics.py [hunk 3]
```diff
@@ -1,25 +1,39 @@
...
+    return 1 - ss_res / (ss_tot + K.epsilon())
 
 
 def mask_coeff_determination(n_channels):
     """split y_true into y_true and mask
 
-    If masked_loss: add a function/method to convert split y_true and pass to
-    loss, metrics and callbacks
-    """
+    For masked_loss there's an added function/method to convert split
```
Comment (anithapk) [inline micro_dl/train/metrics.py]: ~~convert~~

File: micro_dl/train/metrics.py [hunk 4]
```diff
@@ -58,39 +73,101 @@ def dice_coef(y_true, y_pred, smooth=1.):
...
+    :return float min_val: Minimum value of true and pred
+    """
+    min_tensor = tf.stack([K.min(y_true), K.min(y_pred)])
+    max_tensor = tf.stack([K.max(y_true), K.max(y_pred)])
+    min_val = K.min(min_tensor)
+    max_val = K.max(max_tensor) - min_val
+    return max_val, min_val
+
+
+@flip_dimensions
```
Comment (anithapk) [inline micro_dl/train/metrics.py]: 💯

File: micro_dl/train/metrics.py [hunk 5]
```diff
@@ -58,39 +73,101 @@ def dice_coef(y_true, y_pred, smooth=1.):
...
+def ssim(y_true, y_pred):
     """Structural similarity
-    Use max_val=5 to approximate maximum of normalized images
-    tensorflow does not support SSIM for 3D images. Need a different
-    way to compute SSIM for 5D tensor
-    :param tensor y_true: Labeled ground truth
-    :param tensor y_pred: Predicted labels, potentially non-binary
-    :param float max_val: The dynamic range of the images (i.e., the
-        difference between the maximum the and minimum allowed values).
+    Uses a default max_val=6 to approximate maximum of normalized images.
```
Comment (anithapk) [inline micro_dl/train/metrics.py]: max_val here is the dynamic range of input image. B'coz we use z-scored images, setting this to 6 covers 3 standard deviations.
